// @egret/egret-webpack-bundler/lib/loaders/src-loader/Factory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var crypto = __importStar(require("crypto"));
var _fs = __importStar(require("fs"));
var glob = __importStar(require("glob"));
var path = __importStar(require("path"));
var parse_1 = __importDefault(require("./parse"));
var Factory = /** @class */ (function () {
    function Factory(_a) {
        var dirs = _a.dirs, fs = _a.fs;
        this.dirs = dirs;
        this.fs = fs || _fs;
        this.files = {}; // 文件分析缓存
        this.identifiers = {}; // 全部全局变量分布
    }
    Factory.prototype.get = function (fileName) {
        return this.files[fileName] || null;
    };
    Factory.prototype.update = function () {
        var _this = this;
        var files = {};
        this.dirs.forEach(function (dir) {
            glob.sync('**/*.ts', {
                cwd: dir,
            })
                .forEach(function (item) {
                files[path.join(dir, item)] = true;
            });
        });
        Object.keys(files).forEach(function (item) {
            _this.add(item);
        });
        Object.keys(this.files).forEach(function (item) {
            if (!files[item]) {
                // remove file
                _this.remove(item);
            }
        });
    };
    Factory.prototype.remove = function (fileName) {
        var _this = this;
        if (this.files[fileName]) {
            var oldDefines = this.files[fileName].defines;
            // remove identifier
            Object.keys(oldDefines).forEach(function (name) {
                if (_this.identifiers[name]) {
                    _this.identifiers[name].delete(fileName);
                    if (!_this.identifiers[name].size) {
                        delete _this.identifiers[name];
                    }
                }
            });
            delete this.files[fileName];
        }
    };
    Factory.prototype.add = function (fileName) {
        var _this = this;
        var content = this.fs.readFileSync(fileName).toString();
        var hash = crypto.createHash('md5').update(content).digest('hex');
        var files = this.files;
        if (files[fileName] && files[fileName].hash === hash) {
            return;
        }
        this.remove(fileName);
        var _a = parse_1.default(fileName, content), defines = _a.defines, dependencies = _a.dependencies, isModule = _a.isModule;
        // update identifiers
        Object.keys(defines).forEach(function (name) {
            if (!_this.identifiers[name]) {
                _this.identifiers[name] = new Set();
            }
            _this.identifiers[name].add(fileName);
        });
        files[fileName] = {
            hash: hash,
            isModule: isModule,
            dependencies: dependencies,
            defines: defines,
        };
    };
    Factory.prototype.findDependencyFiles = function (dependencies) {
        var _this = this;
        var files = new Set();
        Object.keys(dependencies).forEach(function (key) {
            var thisFiles = null;
            var tmp = key.split('@');
            var names = tmp[0].split('.');
            var namspaces = tmp[1] ? tmp[1].split('.') : [];
            for (var i = namspaces.length; i >= 0; i--) { // 插入一个空的空间
                var ns = namspaces.slice(0, i).join('.');
                for (var j = names.length; j > 0; j--) {
                    var name_1 = names.slice(0, j).join('.');
                    var fullName = (ns ? ns + '.' : '') + name_1;
                    if (_this.identifiers[fullName]) {
                        thisFiles = _this.identifiers[fullName];
                        break;
                    }
                }
                if (thisFiles) {
                    break;
                }
            }
            if (thisFiles) {
                thisFiles.forEach(function (item) {
                    files.add(item);
                });
            }
        });
        return Array.from(files);
    };
    // 排序非模块化文件
    Factory.prototype.sortUnmodules = function () {
        var _this = this;
        var list = Object.keys(this.files)
            .filter(function (file) { return !_this.files[file].isModule; })
            .sort();
        // 冒泡排序
        list.forEach(function (fileName) {
            var dependencyFiles = _this.findDependencyFiles(_this.files[fileName].dependencies);
            var index = list.findIndex(function (item) { return item === fileName; });
            // 筛选在自己后面的文件
            dependencyFiles = dependencyFiles.filter(function (dep) {
                return list.findIndex(function (item) { return item === dep; }) > index;
            });
            if (dependencyFiles.length) {
                list = list.filter(function (item) { return !dependencyFiles.includes(item); });
                list.splice.apply(list, __spreadArrays([index, 0], dependencyFiles));
            }
        });
        return list;
    };
    return Factory;
}());
exports.default = Factory;
