// @egret/egret-webpack-bundler/lib/egretproject/data.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.launcher = exports.projectData = void 0;
var _path = require("path");
var fs = __importStar(require("fs"));
var EgretProjectData = /** @class */ (function () {
    function EgretProjectData() {
        this.egretProperties = {
            modules: [],
            target: { current: "web" },
            engineVersion: "1"
        };
        this.projectRoot = "";
    }
    EgretProjectData.prototype.init = function (projectRoot) {
        this.projectRoot = projectRoot;
        this.reload();
    };
    EgretProjectData.prototype.hasEUI = function () {
        return this.egretProperties.modules.some(function (m) { return m.name == "eui"; });
    };
    EgretProjectData.prototype.reload = function () {
        var egretPropertiesPath = this.getFilePath("egretProperties.json");
        if (fs.existsSync(egretPropertiesPath)) {
            var content = fs.readFileSync(egretPropertiesPath, 'utf-8');
            this.egretProperties = JSON.parse(content);
            var useGUIorEUI = 0;
            for (var _i = 0, _a = this.egretProperties.modules; _i < _a.length; _i++) {
                var m = _a[_i];
                //兼容小写
                if (m.name == "dragonbones") {
                    m.name = "dragonBones";
                }
                if (m.name == "gui" || m.name == "eui") {
                    useGUIorEUI++;
                }
            }
            if (useGUIorEUI >= 2) {
                process.exit(1);
            }
        }
    };
    /**
     * 获取项目的根路径
     */
    EgretProjectData.prototype.getProjectRoot = function () {
        return this.projectRoot;
    };
    EgretProjectData.prototype.getFilePath = function (fileName) {
        return _path.resolve(this.getProjectRoot(), fileName);
    };
    EgretProjectData.prototype.getReleaseRoot = function () {
        var p = "bin-release";
        return p;
        //return file.joinPath(egret.args.projectDir, p);
    };
    EgretProjectData.prototype.getVersionCode = function () {
        return 1;
    };
    EgretProjectData.prototype.getVersion = function () {
        return this.egretProperties.egret_version || this.egretProperties.compilerVersion;
    };
    EgretProjectData.prototype.getIgnorePath = function () {
        return [];
    };
    EgretProjectData.prototype.getCurrentTarget = function () {
        return "web";
        // if (globals.hasKeys(this.egretProperties, ["target", "current"])) {
        //     return this.egretProperties.target.current;
        // }
        // else {
        // }
    };
    EgretProjectData.prototype.getCopyExmlList = function () {
        return [];
    };
    EgretProjectData.prototype.getModulePath2 = function (m) {
        var p = m.path;
        if (!p) {
            var engineVersion = m.version || this.egretProperties.engineVersion;
            var versions = exports.launcher.getEgretToolsInstalledByVersion(engineVersion);
            return _path.join(versions, 'build', m.name);
        }
        var egretLibs;
        if (process.platform === 'linux') {
            egretLibs = _path.resolve(__dirname, '../../');
        }
        else {
            egretLibs = getAppDataEnginesRootPath();
        }
        var keyword = '${EGRET_APP_DATA}';
        if (p.indexOf(keyword) >= 0) {
            p = p.replace(keyword, egretLibs);
        }
        return p;
    };
    EgretProjectData.prototype.getAbsolutePath = function (p) {
        // getAbsolutePath
        if (_path.isAbsolute(p)) {
            return p.split("\\").join("/");
        }
        return _path.join(this.projectRoot, p).split("\\").join("/");
    };
    EgretProjectData.prototype.getModulePath = function (m) {
        var modulePath = this.getModulePath2(m);
        modulePath = this.getAbsolutePath(modulePath);
        var name = m.name;
        if (this.isWasmProject()) {
            if (name == "egret" || name == "eui" || name == "dragonBones" || name == "game") {
                name += "-wasm";
            }
        }
        var searchPaths = [
            _path.join(modulePath, "bin", name),
            _path.join(modulePath, "bin"),
            _path.join(modulePath, "build", name),
            _path.join(modulePath)
        ];
        // if (m.path) {
        //     searchPaths.push(modulePath)
        // }
        if (this.isWasmProject()) {
            searchPaths.unshift(_path.join(modulePath, "bin-wasm"));
            searchPaths.unshift(_path.join(modulePath, "bin-wasm", name));
        }
        var dir = searchPath(searchPaths);
        return dir;
    };
    EgretProjectData.prototype.getLibraryFolder = function () {
        return this.getFilePath('libs/modules');
    };
    EgretProjectData.prototype.getModulesConfig = function (platform) {
        var _this = this;
        if (platform == 'ios' || platform == 'android') {
            platform = 'web';
        }
        var result = this.egretProperties.modules.map(function (m) {
            var name = m.name;
            var sourceDir = _this.getModulePath(m);
            var targetDir = _path.join(_this.getLibraryFolder(), name);
            var relative = _path.relative(_this.getProjectRoot(), sourceDir);
            if (relative.indexOf("..") == -1 && !_path.isAbsolute(relative)) { // source 在项目中
                targetDir = sourceDir;
            }
            targetDir = ((_path.relative(_this.getProjectRoot(), targetDir)) + _path.sep).split("\\").join("/");
            var source = [
                _path.join(sourceDir, name + ".js").split("\\").join("/"),
                _path.join(sourceDir, name + "." + platform + ".js").split("\\").join("/")
            ].filter(fs.existsSync);
            var target = source.map(function (s) {
                var debug = _path.join(targetDir, _path.basename(s)).split("\\").join("/");
                var release = _path.join(targetDir, _path.basename(s, '.js') + '.min.js').split("\\").join("/");
                return {
                    debug: debug,
                    release: release,
                    platform: platform
                };
            });
            return { name: name, target: target, sourceDir: sourceDir, targetDir: targetDir };
        });
        return result;
    };
    EgretProjectData.prototype.isWasmProject = function () {
        return false;
    };
    EgretProjectData.prototype.getResources = function () {
        return ["resource"];
    };
    Object.defineProperty(EgretProjectData.prototype, "useTemplate", {
        get: function () {
            return this.egretProperties.template != undefined;
        },
        enumerable: false,
        configurable: true
    });
    EgretProjectData.prototype.hasModule = function (name) {
        var result = false;
        this.egretProperties.modules.forEach(function (module) {
            if (module.name == name || module.name == name) {
                result = true;
            }
        });
        return result;
    };
    return EgretProjectData;
}());
exports.projectData = new EgretProjectData();
var EgretLauncherProxy = /** @class */ (function () {
    function EgretLauncherProxy() {
    }
    EgretLauncherProxy.prototype.getMinVersion = function () {
        return {
            getAllEngineVersions: '1.0.24',
            getInstalledTools: '1.0.24',
            getTarget: "1.0.45",
            getUserID: "1.0.46",
            sign: "1.0.46"
        };
    };
    EgretLauncherProxy.prototype.getEgretToolsInstalledByVersion = function (checkVersion) {
        if (process.platform === 'linux') {
            return _path.resolve(__dirname, '../../');
        }
        var egretjs = this.getLauncherLibrary();
        var data = egretjs.getAllEngineVersions();
        var versions = [];
        for (var key in data) {
            var item = data[key];
            versions.push({ version: item.version, path: item.root });
        }
        for (var _i = 0, versions_1 = versions; _i < versions_1.length; _i++) {
            var versionInfo = versions_1[_i];
            if (versionInfo.version == checkVersion) {
                return versionInfo.path;
            }
        }
        throw "\u627E\u4E0D\u5230\u6307\u5B9A\u7684 egret \u7248\u672C: " + checkVersion;
    };
    EgretLauncherProxy.prototype.getLauncherLibrary = function () {
        var egretjspath = _path.join(getEgretLauncherPath(), "egret.js");
        var minVersions = this.getMinVersion();
        var m = require(egretjspath);
        var selector = m.selector;
        if (!this.proxy) {
            this.proxy = new Proxy(selector, {
                get: function (target, p, receiver) {
                    var result = target[p];
                    if (!result) {
                        var minVersion = minVersions[p];
                        throw "\u627E\u4E0D\u5230 LauncherAPI:" + p + ",\u8BF7\u5B89\u88C5\u6700\u65B0\u7684\u767D\u9E6D\u5F15\u64CE\u542F\u52A8\u5668\u5BA2\u6237\u7AEF\u89E3\u51B3\u6B64\u95EE\u9898,\u6700\u4F4E\u7248\u672C\u8981\u6C42:" + minVersion + ",\u4E0B\u8F7D\u5730\u5740:https://egret.com/products/engine.html"; //i18n
                    }
                    return result.bind(target);
                }
            });
        }
        return this.proxy;
    };
    return EgretLauncherProxy;
}());
function getAppDataPath() {
    var result = "";
    switch (process.platform) {
        case 'darwin':
            var home = process.env.HOME || ("/Users/" + (process.env.NAME || process.env.LOGNAME));
            if (!home)
                return '';
            result = home + "/Library/Application Support/"; //Egret/engine/`;
            break;
        case 'win32':
            var appdata = process.env.AppData || process.env.USERPROFILE + "/AppData/Roaming/";
            result = appdata.split("\\").join("/");
            break;
        default:
            ;
    }
    if (!fs.existsSync(result)) {
        throw 'missing appdata path';
    }
    return result;
}
function getAppDataEnginesRootPath() {
    var result = _path.join(getAppDataPath(), "Egret/engine/");
    if (!fs.existsSync(result)) {
        throw "\u627E\u4E0D\u5230 " + result + "\uFF0C\u8BF7\u5728 Egret Launcher \u4E2D\u6267\u884C\u4FEE\u590D\u5F15\u64CE"; //todo i18n
    }
    return result;
}
function getEgretLauncherPath() {
    var npmEgretPath;
    if (process.platform === 'darwin') {
        var basicPath = '/usr/local';
        if (!fs.existsSync(basicPath)) { //some mac doesn't have path '/usr/local'
            basicPath = '/usr';
        }
        npmEgretPath = _path.join(basicPath, 'lib/node_modules/egret/EgretEngine');
    }
    else {
        npmEgretPath = _path.join(getAppDataPath(), 'npm/node_modules/egret/EgretEngine');
    }
    if (!fs.existsSync(npmEgretPath)) {
        throw "\u627E\u4E0D\u5230  " + npmEgretPath + "\uFF0C\u8BF7\u5728 Egret Launcher \u4E2D\u6267\u884C\u4FEE\u590D\u5F15\u64CE"; //todo i18n
    }
    var launcherPath = _path.join(fs.readFileSync(npmEgretPath, 'utf-8'), "../");
    return launcherPath;
}
exports.launcher = new EgretLauncherProxy();
function searchPath(searchPaths) {
    for (var _i = 0, searchPaths_1 = searchPaths; _i < searchPaths_1.length; _i++) {
        var s = searchPaths_1[_i];
        if (fs.existsSync(s)) {
            return s;
        }
    }
    return null;
}
function getEgretRoot() {
    var path = require("path");
    var egretRoot;
    var globalpath = module['paths'].concat();
    var existsFlag = false;
    for (var i = 0; i < globalpath.length; i++) {
        var prefix = globalpath[i];
        var url = path.join(prefix, '../');
        if (fs.existsSync(_path.join(url, 'tools/bin/egret'))) {
            existsFlag = true;
            break;
        }
        url = prefix;
        if (fs.existsSync(_path.join(url, 'tools/bin/egret'))) {
            existsFlag = true;
            break;
        }
    }
    if (!existsFlag) {
        throw new Error("can't find Egret");
    }
    egretRoot = url;
    return _path.join(egretRoot, '/').split("\\").join("/");
}
